src/session_protocol.cpp

```diff
@@ -0,0 +1,651 @@
+                proto_proof.rotatingpublickey().size());
+        proof.expiry_unix_ts =
+                std::chrono::sys_seconds(std::chrono::seconds(proto_proof.expiryunixts()));
+        std::memcpy(proof.sig.data(), proto_proof.sig().data(), proto_proof.sig().size());
+
+        if (result.pro_status == ProStatus::Valid) {
+            // Verify the at the proof is verified by the Session Pro Backend key (e.g.: It was
+            // issued by an authoritative backend)
+            if (!proof.verify(pro_backend_pubkey))
+                result.pro_status = ProStatus::InvalidProBackendSig;
+
+            // Check if the proof has expired
+            if (result.pro_status == ProStatus::Valid) {
+                if (unix_ts > result.pro_proof.expiry_unix_ts)
+                    result.pro_status = ProStatus::Expired;
+            }
+        }
+    }
+    return result;
+}
+}  // namespace session
+
+using namespace session;
+
+LIBSESSION_EXPORT
+PRO_FEATURES session_protocol_get_pro_features_for_msg(size_t msg_size, PRO_EXTRA_FEATURES flags) {
+    PRO_FEATURES result = get_pro_features_for_msg(msg_size, flags);
+    return result;
+}
+
+LIBSESSION_EXPORT session_protocol_encrypted_for_destination
+session_protocol_encrypt_for_destination(
+        const void* plaintext,
+        size_t plaintext_len,
+        const void* ed25519_privkey,
+        size_t ed25519_privkey_len,
+        const session_protocol_destination* dest,
+        NAMESPACE space) {
+    session_protocol_encrypted_for_destination result = {};
+    try {
+        EncryptedForDestinationInternal result_internal = encrypt_for_destination_internal(
+                /*plaintext=*/{static_cast<const uint8_t*>(plaintext), plaintext_len},
+                /*ed25519_privkey=*/
+                {static_cast<const uint8_t*>(ed25519_privkey), ed25519_privkey_len},
+                /*dest_type=*/static_cast<DestinationType>(dest->type),
+                /*dest_pro_sig=*/dest->has_pro_sig ? dest->pro_sig : std::span<const uint8_t>(),
+                /*dest_recipient_pubkey=*/dest->recipient_pubkey,
+                /*dest_sent_timestamp_ms=*/std::chrono::milliseconds(dest->sent_timestamp_ms),
+                /*dest_open_group_inbox_server_pubkey=*/dest->open_group_inbox_server_pubkey,
+                /*dest_closed_group_pubkey=*/dest->closed_group_pubkey,
+                /*dest_closed_group_keys=*/
+                static_cast<const config::groups::Keys*>(dest->closed_group_keys->internals),
+                /*space=*/static_cast<config::Namespace>(space),
+                /*use_malloc=*/UseMalloc::Yes);
+
+        result = {
+                .success = true,
+                .encrypted = result_internal.encrypted,
+                .ciphertext = result_internal.ciphertext_c,
+        };
+    } catch (...) {
+    }
+
+    return result;
+}
+
+LIBSESSION_EXPORT
+session_protocol_decrypted_envelope session_protocol_decrypt_envelope(
```

Comment:
How does the memory work for these functions? (ie. is the caller responsible for freeing it?)

I can see that `span_u8_alloc_or_throw` calls `malloc` which I assume will need to be freed somewhere, and I know there is a bunch of complicated malloc'ing in the codebase to simplify how the memory can be freed by the caller (see `make_string_list` in `internal.hpp` and `contacts_blinded` in `contacts.cpp` - there is another in `network_set_paths_changed_callback` in `session_network.cpp` but I've got a refactored version locally which behaves more similarly to the others)

Reply:
Apologies, yes I've documented these allocations now. The spans are allocated with the C runtime's (CRT) `malloc` so it needs to be freed with the equivalent `free` on the `data` member.

Reply:
Just wondering whether we should add `session_protocol_decrypted_envelope_free`/`session_protocol_encrypted_for_destination_free` functions in case the structures change in the future?

That way the native code just needs to call `result = session_protocol_decrypt_envelope(...)` and `session_protocol_decrypted_envelope_free(result)` and they can trust that the data will be freed correctly, whereas if the native code is responsible for freeing individual variables of the returned object then it'd be pretty easy to miss a change that results in a memory leak